agent/internal/fluent/fluent.go

```diff
@@ -0,0 +1,502 @@
+package fluent
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/docker/docker/pkg/stdcopy"
+
+	"github.com/determined-ai/determined/agent/internal/container"
+	"github.com/determined-ai/determined/agent/internal/options"
+	"github.com/determined-ai/determined/agent/pkg/docker"
+	"github.com/determined-ai/determined/master/pkg/aproto"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/syncx/errgroupx"
+
+	"github.com/docker/docker/api/types"
+	dcontainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/master/pkg/fluent"
+)
+
+// EnvVarNames are the of environment variables whose values should be included in log entries that
+// Docker or the agent sends to the Fluent Bit logger.
+var EnvVarNames = []string{
+	container.ContainerIDEnvVar,
+	container.AllocationIDEnvVar,
+	container.TaskIDEnvVar,
+}
+
+var fluentLogLineRegexp = regexp.MustCompile(`\[[^]]*\] \[ *([^]]*)\] (.*)`)
+
+// Fluent manages the lifecycle of the Fluent Bit container that is run by the agent for
+// the purpose of forwarding container logs.
+type Fluent struct {
+	// Configuration details.
+	opts  options.AgentOptions
+	mopts aproto.MasterSetAgentOptions
+
+	// System dependencies.
+	log    *log.Entry
+	docker *docker.Client
+
+	// Internal state.
```

Comment:
`// Circular buffer used for printing recent logs on fluent container failures.`

Reply:
in fact it seems like this could have been `container/ring` like the one used in container manager

Reply:
yeah, didn't change this but i agree, could be. we should just write a generic ring - we already have it in a few places.